src/foundry/common/abstract/type-data.d.mts

```diff
@@ -19,9 +22,53 @@ declare class _InternalTypeDataModel<
   BaseData extends Record<string, unknown> = Record<string, never>,
   DerivedData extends Record<string, unknown> = Record<string, never>,
   // This does not work if inlined. It's weird to put it here but it works.
-  _ComputedInstance extends object = Merge<BaseData, DerivedData>,
+  _ComputedInstance extends object = RemoveIndexSignatures<Merge<BaseData, DerivedData>>,
 > extends _InternalTypeDataModelConst<Schema, Parent, _ComputedInstance> {}
 
+// These properties are used to give a performant way of inferring the `BaseModel`, `BaseData` and `DerivedData` types.
+// This avoids checking the entire constraint `T extends TypeDataModel.Any` to infer out the `BaseData` and `DerivedData` types.
+// To prevent adding properties that could appear at runtime they are unique symbols.
+// This means that the only way to access them is to have a reference to the variables which will be impossible outside of this file.
+declare const __BaseModel: unique symbol;
+declare const __BaseData: unique symbol;
+declare const __DerivedData: unique symbol;
+
+declare namespace TypeDataModel {
+  export type Any = TypeDataModel<any, any, any, any>;
+
+  // The only way to have these properties is to be descended from `TypeDataField` because these unique symbols aren't used elsewhere.
+  type TypeDataModelInternal<BaseModel, BaseData, DerivedData> = {
+    [__BaseModel]: BaseModel;
+    [__BaseData]: BaseData;
+    [__DerivedData]: DerivedData;
+  };
```

Comment:
Especially when dealing with `this` you need to be careful with complex bounds. `this` is notorious for this because it's inherently cyclical and TypeScript doesn't have good proofs around this so it literally just loops checking the same constraint over and over in the presence of `this`. 

I may consider using this strategy elsewhere. It's ugly but as far as I can tell it's a genuine performance win.

In this case specifically there were problems where the type for something like `this[key]` wouldn't reduce down from `this[key]`; it knew the key existed but it didn't give a value, even `any`, and so broke every test. I could probably remove this and still have the type work now but the performance wins could be nice.